### DISCUSSION TITLE
Floats instead of integer for lat/lon settings in xarray coarsen function

### DISCUSSION THREAD

**Discussion by jwynnsmith** [pydata/xarray#6266] [Q&A]
Hello

I will compare a downscaled (upsampled) goes16 grid to a coarse 1.25 degree lon by 1 degree lat model grid.

To match the precise collocated obs vs model grids I will need to use floats for the lat/lon settings in xarray coarsen.  When I do that I get the error below that basically says lon and lat have to be integers. Is there any way to override or ignore this error to get the function coarsen function to calculate with the float datatype? (i.e. try except [did not work for me but maybe I have commands out of order])

Error message here:
glm_downscale = ds.coarsen(lon=41.75, lat=33.42, boundary=‘pad’).mean()
File “/app/spack/v0.15/linux-rhel7-x86_64/gcc-4.8.5/python/3.7.7-d6cyi6ophaei6arnmzya2kn6yumye2yl/lib/python3.7/site-packages/xarray/core/rolling.py”, line 1036, in wrapped_func
**kwargs,
File “/app/spack/v0.15/linux-rhel7-x86_64/gcc-4.8.5/python/3.7.7-d6cyi6ophaei6arnmzya2kn6yumye2yl/lib/python3.7/site-packages/xarray/core/variable.py”, line 2194, in coarsen
reshaped, axes = self.coarsen_reshape(windows, boundary, side)
File “/app/spack/v0.15/linux-rhel7-x86_64/gcc-4.8.5/python/3.7.7-d6cyi6ophaei6arnmzya2kn6yumye2yl/lib/python3.7/site-packages/xarray/core/variable.py”, line 2248, in coarsen_reshape
variable = variable.pad(pad_width, mode=“constant”)
File “/app/spack/v0.15/linux-rhel7-x86_64/gcc-4.8.5/python/3.7.7-d6cyi6ophaei6arnmzya2kn6yumye2yl/lib/python3.7/site-packages/xarray/core/variable.py”, line 1350, in pad
**pad_option_kwargs,
File “/app/spack/v0.15/linux-rhel7-x86_64/gcc-4.8.5/python/3.7.7-d6cyi6ophaei6arnmzya2kn6yumye2yl/lib/python3.7/site-packages/xarray/core/duck_array_ops.py”, line 56, in f
return wrapped(*args, **kwargs)
File “<array_function internals>”, line 6, in pad
File “/home/Jonathan.Smith/.local/lib/python3.7/site-packages/numpy/lib/arraypad.py”, line 740, in pad
raise TypeError(’pad_width must be of integral type.’)
TypeError: pad_width must be of integral type.

Thanks,
Jonathan

**Comment by mathause** [comment]
`coarsen` can only be used to average over full grid points and is therefore not the tool for what you need. You will have to use something like [cdo](https://code.mpimet.mpg.de/projects/cdo/) (which also has [python bindings](https://code.mpimet.mpg.de/projects/cdo/wiki/Cdo%7Brbpy%7D)) or [xESMF](https://xesmf.readthedocs.io/en/latest/).

---

I personally mostly use cdo, even though it is a bit more clumsy to use, because it supports more grids and regridding schemes, but depending on your use case xESMF might be enough for you.

I add an example how to use cdo from python because I find it non-trivial to do (it reads and writes a netCDF file from and to disk):

```python
import cdo

CDO = cdo.Cdo()

CDO.remapcon2(
    f'{target_grid}.txt', options='-b F64', input=load_filename, output=save_filename
    )
```

**Reply by jwynnsmith** [reply]
Thanks for your reply @mathause 
I will move forward with the cdo command above.  I have tried remapbil and other cdo commands to no avail.

Also, I worked with xesmf
After forming a netcdf file I used the commands as instructed for curvilinear grids on this page: https://xesmf.readthedocs.io/en/latest/notebooks/Curvilinear_grid.html

Here are the commands:

```python
ds_out = xe.util.grid_global(1.25, 1)

regridder = xe.Regridder(ds, ds_out, "bilinear", periodic=True)
glm_ff_downscale = regridder(glm_ff)
```

And obtained this error:
```python-traceback
Traceback (most recent call last):
  File "lat_lon_xesmf2.py", line 60, in <module>
    regridder = xe.Regridder(ds, ds_out, "bilinear")
  File "/app/spack/v0.15/linux-rhel7-x86_64/gcc-4.8.5/python/3.7.7-d6cyi6ophaei6arnmzya2kn6yumye2yl/lib/python3.7/site-packages/xesmf/frontend.py", line 772, in __init__
    super().__init__(grid_in, grid_out, method, input_dims=input_dims, **kwargs)
  File "/app/spack/v0.15/linux-rhel7-x86_64/gcc-4.8.5/python/3.7.7-d6cyi6ophaei6arnmzya2kn6yumye2yl/lib/python3.7/site-packages/xesmf/frontend.py", line 315, in __init__
    weights = self._compute_weights()  # Dictionary of weights
  File "/app/spack/v0.15/linux-rhel7-x86_64/gcc-4.8.5/python/3.7.7-d6cyi6ophaei6arnmzya2kn6yumye2yl/lib/python3.7/site-packages/xesmf/frontend.py", line 375, in _compute_weights
    ignore_degenerate=self.ignore_degenerate,
  File "/app/spack/v0.15/linux-rhel7-x86_64/gcc-4.8.5/python/3.7.7-d6cyi6ophaei6arnmzya2kn6yumye2yl/lib/python3.7/site-packages/xesmf/backend.py", line 481, in esmf_regrid_build
    regrid = ESMF.Regrid(sourcefield, destfield, **kwargs)
  File "/app/spack/v0.15/linux-rhel7-x86_64/gcc-4.8.5/python/3.7.7-d6cyi6ophaei6arnmzya2kn6yumye2yl/lib/python3.7/site-packages/ESMF/util/decorators.py", line 64, in new_func
    return func(*args, **kwargs)
  File "/app/spack/v0.15/linux-rhel7-x86_64/gcc-4.8.5/python/3.7.7-d6cyi6ophaei6arnmzya2kn6yumye2yl/lib/python3.7/site-packages/ESMF/api/regrid.py", line 185, in __init__
    dstFracField=dst_frac_field
  File "/app/spack/v0.15/linux-rhel7-x86_64/gcc-4.8.5/python/3.7.7-d6cyi6ophaei6arnmzya2kn6yumye2yl/lib/python3.7/site-packages/ESMF/interface/cbindings.py", line 2068, in ESMP_FieldRegridStore
    handle_esmf_error(rc, 'ESMC_FieldRegridStore')
  File "/app/spack/v0.15/linux-rhel7-x86_64/gcc-4.8.5/python/3.7.7-d6cyi6ophaei6arnmzya2kn6yumye2yl/lib/python3.7/site-packages/ESMF/interface/cbindings.py", line 26, in handle_esmf_error
    raise ValueError(msg)
ValueError: ESMC_FieldRegridStore failed with rc = 506. Please check the log files (named "*ESMF_LogFile").
```

I looked in the error log file listed above and did not know what to do with the message listed below:
```
20220208 021505.857 ERROR            PET0 ESMF_Regrid.F90:366 ESMF_RegridStore Invalid argument  - Internal subroutine call returned Error
20220208 021507.897 ERROR            PET0 ESMF_FieldRegrid.F90:1415 ESMF_FieldRegridStoreNX Invalid argument  - Internal subroutine call returned Error
20220208 021508.077 ERROR            PET0 ESMF_Field_C.F90:1130 f_esmf_regridstore Invalid argument  - Internal subroutine call returned Error
20220208 021543.915 INFO             PET0 Finalizing ESMF
```

Any help in the interpretation of this error will be greatly appreciated.

Thanks
Jonathan
